### PR TITLE
Get subdir from registry.

### PR DESCRIPTION
**Pull request by GunnarFarneback**
Followup to #1422 and https://github.com/JuliaRegistries/RegistryTools.jl/pull/31 to read `subdir` information from registry.

Demo examples to follow after @DilumAluthge merges https://github.com/DilumAluthge/TestRegistry/pull/1. :-)

### CODE REVIEW

**Comment by GunnarFarneback** [comment]
```
$ ~/.julia/dev/Pkg$ mkdir /tmp/testdepot
$ ~/.julia/dev/Pkg$ mkdir /tmp/testproject
$ ~/.julia/dev/Pkg$ JULIA_DEPOT_PATH=/tmp/testdepot ~/julia1.4/julia --project
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.4.0 (2020-03-21)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> import Pkg
[ Info: Precompiling Pkg [5f3860df-0467-5c1b-a2e3-6b6828f907f9]

(Pkg) pkg> activate /tmp/testproject
 Activating new environment at `/tmp/testproject/Project.toml`

(testproject) pkg> registry add General
    Cloning registry from "https://github.com/JuliaRegistries/General.git"
      Added registry `General` to `/tmp/testdepot/registries/General`

(testproject) pkg> registry add https://github.com/DilumAluthge/TestRegistry.git
    Cloning registry from "https://github.com/DilumAluthge/TestRegistry.git"
      Added registry `TestRegistry` to `/tmp/testdepot/registries/TestRegistry`

(testproject) pkg> develop Foo
    Cloning git-repo `https://github.com/DilumAluthge/FooBar.git`
  Resolving package versions...
Updating `/tmp/testproject/Project.toml`
  [ac29586c] + Foo v0.1.0 `/tmp/testdepot/dev/FooBar.git/myfolder/Foo`
Updating `/tmp/testproject/Manifest.toml`
  [ac29586c] + Foo v0.1.0 `/tmp/testdepot/dev/FooBar.git/myfolder/Foo`

(testproject) pkg> status
Status `/tmp/testproject/Project.toml`
  [ac29586c] Foo v0.1.0 `/tmp/testdepot/dev/FooBar.git/myfolder/Foo`

(testproject) pkg> test Foo
    Testing Foo
Status `/tmp/jl_J4KKbT/Project.toml`
  [ac29586c] Foo v0.1.0 `/tmp/testdepot/dev/FooBar.git/myfolder/Foo`
  [8dfed614] Test
Status `/tmp/jl_J4KKbT/Manifest.toml`
  [ac29586c] Foo v0.1.0 `/tmp/testdepot/dev/FooBar.git/myfolder/Foo`
[...]

Test Summary: | Pass  Total
Foo.jl        |    1      1
    Testing Foo tests passed 

(testproject) pkg> develop Bar
Path `/tmp/testdepot/dev/FooBar.git` exists and looks like the correct repo. Using existing path.
  Resolving package versions...
[...]
Updating `/tmp/testproject/Project.toml`
  [f73876a8] + Bar v0.1.0 `/tmp/testdepot/dev/FooBar.git/otherfolder/subfolder/Bar`
Updating `/tmp/testproject/Manifest.toml`
  [f73876a8] + Bar v0.1.0 `/tmp/testdepot/dev/FooBar.git/otherfolder/subfolder/Bar`
[...]

(testproject) pkg> status
Status `/tmp/testproject/Project.toml`
  [f73876a8] Bar v0.1.0 `/tmp/testdepot/dev/FooBar.git/otherfolder/subfolder/Bar`
  [ac29586c] Foo v0.1.0 `/tmp/testdepot/dev/FooBar.git/myfolder/Foo`

(testproject) pkg> test Bar
    Testing Bar
Status `/tmp/jl_pge8fD/Project.toml`
  [f73876a8] Bar v0.1.0 `/tmp/testdepot/dev/FooBar.git/otherfolder/subfolder/Bar`
  [a93c6f00] DataFrames v0.18.4
  [8dfed614] Test
Status `/tmp/jl_pge8fD/Manifest.toml`
  [f73876a8] Bar v0.1.0 `/tmp/testdepot/dev/FooBar.git/otherfolder/subfolder/Bar`
[...]
Test Summary: | Pass  Total
Bar.jl        |    2      2
    Testing Bar tests passed 

(testproject) pkg> rm Bar
Updating `/tmp/testproject/Project.toml`
  [f73876a8] - Bar v0.1.0 `/tmp/testdepot/dev/FooBar.git/otherfolder/subfolder/Bar`
Updating `/tmp/testproject/Manifest.toml`
  [f73876a8] - Bar v0.1.0 `/tmp/testdepot/dev/FooBar.git/otherfolder/subfolder/Bar`
[...]

(testproject) pkg> add Bar#master
    Cloning git-repo `https://github.com/DilumAluthge/FooBar.git`
   Updating git-repo `https://github.com/DilumAluthge/FooBar.git`
   Updating registry at `/tmp/testdepot/registries/General`
   Updating git-repo `https://github.com/JuliaRegistries/General.git`
   Updating registry at `/tmp/testdepot/registries/TestRegistry`
   Updating git-repo `https://github.com/DilumAluthge/TestRegistry.git`
  Resolving package versions...
Updating `/tmp/testproject/Project.toml`
  [f73876a8] + Bar v0.1.0 `https://github.com/DilumAluthge/FooBar.git/[otherfolder/subfolder/Bar]#master`                                                                             
Updating `/tmp/testproject/Manifest.toml`
  [f73876a8] + Bar v0.1.0 `https://github.com/DilumAluthge/FooBar.git/[otherfolder/subfolder/Bar]#master`
[...]

(testproject) pkg> status
Status `/tmp/testproject/Project.toml`
  [f73876a8] Bar v0.1.0 `https://github.com/DilumAluthge/FooBar.git/[otherfolder/subfolder/Bar]#master`
  [ac29586c] Foo v0.1.0 `/tmp/testdepot/dev/FooBar.git/myfolder/Foo`

(testproject) pkg> test Bar
    Testing Bar
Status `/tmp/jl_yhqxpP/Project.toml`
  [f73876a8] Bar v0.1.0 `https://github.com/DilumAluthge/FooBar.git/[otherfolder/subfolder/Bar]#master`
  [a93c6f00] DataFrames v0.18.4
  [8dfed614] Test
Status `/tmp/jl_yhqxpP/Manifest.toml`
  [f73876a8] Bar v0.1.0 `https://github.com/DilumAluthge/FooBar.git/[otherfolder/subfolder/Bar]#master`
[...]
Test Summary: | Pass  Total
Bar.jl        |    2      2
    Testing Bar tests passed 
```

**Comment by GunnarFarneback** [comment]
From the size of this patch it's easily deduced that @KristofferC did all the heavy lifting in #1422.
This covers a good part of #1251.

File: src/Types.jl [hunk 1]
```diff
@@ -539,6 +539,11 @@ function set_repo_source_from_registry!(ctx, pkg)
     end
     _, repo_source = repo_info[1] # Just take the first repo we found
     pkg.repo.source = repo_source
+    subdir_info = registered_info(ctx, pkg.uuid, "subdir")
+    _, subdir = subdir_info[1] # Just take the first repo we found
```
Comment (KristofferC) [inline src/Types.jl:543]:
```suggestion
    _, subdir = subdir_info[1] # Just take the first subdir we found
```

?

Comment (GunnarFarneback) [reply src/Types.jl:543]: Yeah, copy/paste error.

**Comment by KristofferC** [comment]
Can we have a test using TestRegistry?

**Comment by GunnarFarneback** [comment]
I'm working on tests but I'd prefer to make them independent of external repositories. It may take a few days though.

**Comment by GunnarFarneback** [comment]
Well, that didn't go too well in CI. I managed to get the tests to pass locally with `include("test/runtests.jl")`, but I don't know if that's the right way to do it. Any ideas what should be done differently in the tests?

**Comment by GunnarFarneback** [comment]
So the windows failure is, not too surprisingly, to start with a question of the direction of slashes and building a file URL.

On Linux it has something to do with interaction with other test files. It fails locally too if I don't comment out other tests.

**Comment by GunnarFarneback** [comment]
One of the Travis builds passed now, the other three, as well as AppVeyor, have download failures:
```
ERROR: LoadError: LoadError: could not download https://pkg.julialang.org/registry/23338594-aafe-5451-b93e-139f81909106/5ad52a12891bd9dbc19543f34b1c24a9d72495a3
```
which seem to happen intermittently. I don't think it's possible to get full green while the download problems remain, so please review.

**Comment by DilumAluthge** [comment]
Add these lines to the `.travis.yml` file:
```yaml
env:
  global:
    - JULIA_PKG_SERVER=""
```

**Comment by DilumAluthge** [comment]
For AppVeyor, try editing lines 39 to 41 of `appveyor.yml` so they look like this:
```yaml
test_script:
  - echo "%JL_TEST_SCRIPT%"
  - C:\julia\bin\julia -e "ENV[string(:JULIA_PKG_SERVER)] = string(); %JL_TEST_SCRIPT%"
```

**Comment by GunnarFarneback** [comment]
Disabling the PKG SERVER was effective to get the tests to pass. All is green now.

**Comment by DilumAluthge** [comment]
Now that CI is passing, it would be great if @KristofferC could do one final review before merging.

File: .travis.yml [hunk 2]
```diff
@@ -9,6 +9,10 @@ os:
   - linux
   - osx
 
+env:
+  global:
+    - JULIA_PKG_SERVER=""
```
Comment (KristofferC) [inline .travis.yml:14]: Adding this globally feels like a bit of a problem since the whole Julia ecosystem is running with this on (in 1.5).

Comment (GunnarFarneback) [reply .travis.yml:14]: I don't have a strong opinion about this other than that it prevented meaningful CI results at the time. I can back it out.

Comment (GunnarFarneback) [reply .travis.yml]: Very funny. With a `git reset`, Travis just reported the old results for that commit.

**Comment by GunnarFarneback** [comment]
Pkg server registry download failure on AppVeyor. Artifact download failure on Travis Mac. Neither related to this PR.

**Comment by KristofferC** [comment]
Let's get this in and we can work on CI stability separately.

**Pull request merged**